File: Dynamic Programming/Medium/LET_no647_Palindromic Substrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        is_palindrome = [[False for _ in range(n)] for _ in range(n)]
        
        res = 0
        for d in range(n):
            for left in range(n - d):
                right = left + d
                is_palindrome[left][right] = s[left] == s[right] and (right - left < 3 or is_palindrome[left + 1][right - 1])
                res += is_palindrome[left][right]
        
        return res

File: Dynamic Programming/Medium/LET_no647_Palindromic Substrings.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size();
        vector<vector<bool>> isPalindrome(n, vector<bool>(n, false));

        int res = 0;
        for (int d = 0; d < n; d++) {
            for (int left = 0; left < n - d; left++) {
                int right = left + d;
                isPalindrome[left][right] = s[left] == s[right] && (right - left < 3 || isPalindrome[left + 1][right - 1]);
                res += isPalindrome[left][right];
            }
        }

        return res;
    }
};